### DISCUSSION TITLE
pattern_state_machine

### DISCUSSION THREAD

**Discussion by giscus** [ruanqizhen/labview_book#64] [Announcements]
# 状态机

https://labview.qizhen.xyz/pattern_state_machine/

**Comment by zhangyongzhoueric** [comment]
讲的真好

**Comment by Dadaer123456** [comment]
labview2018做最后程序的时候，"int"状态下的字符数组队列预览条件[“0”.."9"]遇到数字9时无法计算，改为["0".."9","9"]就好了，不知高版本也有同样问题没有？

**Reply by ruanqizhen** [reply]
你说的没错，我之前都没有注意到过。LabVIEW 的选择框，如果输入条件是字符，某个条件覆盖了一段字符的话，比如 [“0”.."9"]，最后一个字符"9"是无法落入这个分支的。如果需要让"9"也被选中，条件就要设置到"9"之后的那个字符 ":"
这个行为与其它数据类型的选择条件不一致，比如条件如果是一段数字 [0..9]， 那么数字 9 是会落入这个分支的。LabVIEW 各个版本都如此，估计也不会改了。我会在书里更新说明一下这个问题。

**Comment by zsl-nuaa** [comment]
老师你好 最后的练习我有几个问题
1.我放事件1里黄灯的闪烁为什么在其他事件里面把黄灯的局部变量F后(这不是让黄灯直接熄灭吗)还会继续闪 得把闪烁放每一个事件然后都F后才不会闪
2.放循环的停止键感觉有时没反应 循环完成也不停止（有时会停），就是放探针一直按停止键看到布尔值也没变。另外我想实现的是和中止程序一样的功能，停止键能打断循环然后所有灯灭，该怎么做呢 谢谢
https://github.com/zsl-nuaa/labview-practise/blob/main/test2.png

**Reply by ruanqizhen** [reply]
1. 我看了一下你的程序，它在选择结构里，先执行 1 分支，然后会执行 2 分支。如果在 2 分支中为黄灯闪烁属性输入一个 F 值，它是可以明显看到停止闪烁的。在其它分支中停止闪烁也没问题，只不过停下的时间短，你可能没有明显观察到。
2. 因为循环中带有长延时，程序经常无法及时处理停止键的按下。要让停止键稳定工作，需要使用事件结构。

**Reply by zsl-nuaa** [reply]
请问 那怎么能让事件结构里的while循环停止呢 程序运行点开始后 停止键按不下去

**Reply by ruanqizhen** [reply]
通常，应该把事件结构放在循环结构的里面，这样可以比较直接的控制循环停下。
如果一定要把某个循环，或者运行时间比较长的代码段放在事件结构的一个分支中，那么需要多线程编程的机制来控制这部分代码。界面部分，和循环部分要运行在不同的线程中，这样才能运行循环的同时，也接受到用户的按键操作。详细的方法可以参考 https://lv.qizhen.xyz/vi_server_for_subvi 中介绍的内容。
